x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/find_rules_route.ts

```diff
@@ -113,3 +165,94 @@ export const findRulesRoute = (router: SecuritySolutionPluginRouter) => {
     }
   );
 };
+
+// const bulkFetchRuleActionsOfManyRules = async (
+//   client: ElasticsearchClient,
+//   ruleIds: string[]
+// ): Promise<Array<RulesActionsSavedObject | null>> => {
+//   if (ruleIds.length === 0) {
+//     return [];
+//   }
+
+//   const response = await client.search({
+//     index: '.kibana',
+//     ignore_unavailable: true,
+//     size: ruleIds.length,
+//     body: {
+//       size: ruleIds.length,
+//       query: {
+//         bool: {
+//           filter: [
+//             { term: { type: 'siem-detection-engine-rule-actions' } },
+//             { terms: { 'siem-detection-engine-rule-actions.ruleAlertId': ruleIds } },
+//           ],
+//         },
+//       },
+//     },
+//   });
+
+//   const actionsHits = response.body.hits?.hits ?? [];
+//   const actions = actionsHits.filter(Boolean).map((hit: any) => {
+//     const id = hit._id;
+//     const attributes = hit._source?.['siem-detection-engine-rule-actions'] ?? {};
+//     return { id, ...attributes };
+//   });
+
+//   return actions.filter(Boolean) as RulesActionsSavedObject[];
+// };
+
+const bulkFetchLatestStatusesOfManyRules = async (
+  client: ElasticsearchClient,
+  ruleIds: string[]
+): Promise<IRuleSavedAttributesSavedObjectAttributes[]> => {
+  // This is not only a performance optimization. If ids is empty, the search
+  // request to Elasticsearch using terms aggregation will throw a 400 error,
+  // which is not what we want here.
+  if (ruleIds.length === 0) {
+    return [];
+  }
+
+  const response = await client.search({
+    index: '.kibana',
+    ignore_unavailable: true,
+    size: 0,
+    body: {
+      size: 0,
+      query: {
+        bool: {
+          filter: [
+            { term: { type: 'siem-detection-engine-rule-status' } },
+            { terms: { 'siem-detection-engine-rule-status.alertId': ruleIds } },
+          ],
+        },
+      },
+      aggs: {
+        rules: {
+          terms: {
+            field: 'siem-detection-engine-rule-status.alertId',
+            size: ruleIds.length,
+          },
+          aggs: {
+            latest_status: {
+              top_hits: {
+                size: 1,
+                sort: [{ 'siem-detection-engine-rule-status.statusDate': { order: 'desc' } }],
+              },
+            },
+          },
+        },
+      },
+    },
+  });
+
+  const ruleBuckets = response.body.aggregations?.rules?.buckets ?? [];
+  const statusHits = ruleBuckets.map((bucket: any) => {
+    const [firstHit] = bucket?.latest_status?.hits?.hits ?? [];
+    return firstHit ?? null;
+  });
+  const statuses = statusHits.map((hit: any) => {
+    return hit?._source?.['siem-detection-engine-rule-status'] ?? null;
+  });
+
+  return statuses.filter(Boolean) as IRuleSavedAttributesSavedObjectAttributes[];
+};
```

Comment:
Please review this from the security standpoint (access privileges, isolation (spaces), error handling, potential issues, etc.).

Note that `ruleIds` here are not coming from request parameters. We retrieve them via alerting client which uses saved objects client. Which should handle spaces and RBAC for us correctly.

Our "rule actions" and "rule statuses" have "foreign keys" pointing to detections rules via their ids.

So if a rule (saved object with type `alert`) has an `"_id": "alert:b921e020-6a13-11eb-bf61-5f56ac457c9b"`, its status object will contain a foreign key `"alertId": "b921e020-6a13-11eb-bf61-5f56ac457c9b"` in its saved object attributes.